sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/SpecificParquetRecordReaderBase.java

```diff
@@ -103,9 +104,14 @@ public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptCont
     fileReader.setRequestedSchema(requestedSchema);
     String sparkRequestedSchemaString =
         configuration.get(ParquetReadSupport$.MODULE$.SPARK_ROW_REQUESTED_SCHEMA());
-    this.sparkSchema = StructType$.MODULE$.fromString(sparkRequestedSchemaString);
+    StructType sparkRequestedSchema = StructType$.MODULE$.fromString(sparkRequestedSchemaString);
+    ParquetToSparkSchemaConverter converter = new ParquetToSparkSchemaConverter(configuration);
+    this.parquetColumn = converter.convertParquetColumn(requestedSchema,
```

Comment:
This line has a performance impact on parquet files with very wide records.

Take, for example, with a parquet dataset with:

- 6000 bigint columns
- 25 files
- 5000 records per file (total of 125000 records)

Reading all the records in this parquet dataset with a single executor thread is 42% slower with this PR than with the baseline (commit f361ad8d2e on the master branch).

The main culprit is the n**2 issue [here](https://github.com/apache/spark/blob/f361ad8d2eb067ec86483a294771f1976f31a981/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala#L95)  in ParquetToSparkSchemaConverter#convertInternal, which was not exercised anywhere except by unit tests until your PR (as far as I can tell).

By the way, even with a patch for the above n**2 issue, this PR is still 10% slower than the baseline when testing the wide row case. I haven't confirmed it, but I think this is due to the extra OnHeapColumnVector instances created in ParquetColumnVector, for which you already have a TODO.

Reply:
Thanks! let me try to reproduce the above scenario (it'd be great if you already have some code snippet to share) locally and see what's the causing the perf issue in `ParquetToSparkSchemaConverter#convertInternal`. Ideally we can skip the code path if the schema doesn't contain complex types.

Reply:
>see what's the causing the perf issue in ParquetToSparkSchemaConverter#convertInternal

It's definitely the n\*n (or n\*(n/2) on average) loop here:
```
    val converted = (0 until groupColumn.getChildrenCount).map { i =>
      val field = groupColumn.getChild(i)
      val fieldFromReadSchema = sparkReadSchema.flatMap { schema =>
        schema.find(f => isSameFieldName(f.name, field.getName, caseSensitive))
      }
```

`schema.find` should instead be a map lookup, or something like that.

>it'd be great if you already have some code snippet to share

Sure, in a verbose state:

```
// test for wide rows

val home_candidate = sys.env("HOME")
val home = if (!home_candidate.endsWith("/")) {
  s"${home_candidate}/"
} else {
  home_candidate
}

val width = 6000

val selectExpr = (1 to width).map { i =>
  s"value as c$i"
}


import scala.util.Random
val r = new Random(65657652L)

val loopCount = 25
for (i <- 0 until loopCount) {
  // println(s"iteration $i")
  val df = spark.range(5000).map(_ => r.nextLong).toDF().selectExpr(selectExpr: _*)
  val mode = if (i == 0) {
    "overwrite"
  } else {
    "append"
  }
  df.coalesce(1).write.mode(mode).format("parquet").save(s"${home}data/wide_row_parquet")
}


// read test

val home_candidate = sys.env("HOME")
val home = if (!home_candidate.endsWith("/")) {
  s"${home_candidate}/"
} else {
  home_candidate
}

spark.read.parquet(s"${home}data/wide_row_parquet").createOrReplaceTempView("df")
val startTime = System.currentTimeMillis
sql("select * from df where (c5*c12) == 12").collect
(System.currentTimeMillis - startTime)/60.0/1000
```

Reply:
Thanks again @bersprockets ! let me address the issue and update the PR.

Reply:
Updated the PR. I also re-run the above test which used to take 1.05m before the fix, and improved to 0.58m afterwards, which is very close to the number in master branch. 

@bersprockets it'd be great if you can verify it too. Thanks.

Reply:
Yes, you've eliminated the schema lookup penalty.

There still seems to be the remaining 10-15% performance penalty. Earlier, I speculated on the cause of this particular penalty:

>this PR is still 10% slower than the baseline when testing the wide row case. I haven't confirmed it, but I think this is due to the extra OnHeapColumnVector instances created in ParquetColumnVector, for which you already have a TODO.

So maybe that will get resolved once the TODO is resolved.

Reply:
Thanks, yes I'm aware of the other one: we should not need to allocate the vectors for definition & repetition levels when the schema is flat. I'm hoping to address this separately with another PR though - don't want to make this one too bloated :)

BTW @bersprockets : how can I reproduce the 10-15% performance penalty? I was using the above code snippet and got almost the same numbers on my machine with the latest fix.

Reply:
>how can I reproduce the 10-15% performance penalty

Maybe my old 2015 2.5 GHz MBP just feels it more strongly.

I launched spark-shell it with this command line
```
bin/spark-shell --driver-memory 5g --master "local[1]"
```
Maybe the 5g makes a difference (e.g., more pressure on the garbage collector when there are lots of OnHeapColumnVector instances).

Reply:
I have the follow-up to skip creating these column vectors and reading definition levels if unnecessary done [here](https://github.com/sunchao/spark/commit/f968791635db1d679638511fc4ff01d2b1eccff0). I can include it to the PR if it doesn't complicated the review :) cc @sadikovi